src/util/timer.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::time::Duration;
+use util::time::Instant;
+use std::collections::BinaryHeap;
+
+pub struct TimeoutTask<T> {
+    next_tick: Instant,
+    timeout: Duration,
+    task: T,
+}
+
+impl<T> TimeoutTask<T> {
+    pub fn restore(self, timer: &mut Timer<T>) {
```

Comment:
What does it mean?

Reply:
After we got tasks, they are removed from `Timer`. So call `restore` on it can add it back.

Reply:
Does `Timer` trace historical tasks? Is there any difference between a new task and a historical task?

Reply:
`Timer` can trace it. So when we `restore` it, `Timer` knows when it was triggered last time.